File: apps/web/src/lib/reviews.ts
"use client";

/** Review requests: sending them, answering them, and the inbox.
 *
 * Everything here is addressed by *nickname* when a human types it and
 * by id everywhere else. The backend re-checks who the caller is on
 * every call, so nothing in this file is a permission decision — it
 * only decides what to show.
 */

import { authFetch } from "./auth";

export type ReviewStage = "spec" | "result";
export type ReviewStatus = "pending" | "approved" | "rejected" | "cancelled";

export interface ReviewRequest {
  id: string;
  benchmark_id: string;
  stage: ReviewStage;
  requested_by_user_id: string;
  reviewer_user_id: string;
  status: ReviewStatus;
  request_comment: string;
  review_comment: string;
  created_at: string;
  reviewed_at: string | null;
  cancelled_at: string | null;
}

export interface UserSummary {
  id: string;
  nickname: string;
  display_name: string;
  avatar_url: string;
}

export interface ReviewRequestView {
  request: ReviewRequest;
  benchmark_name: string;
  benchmark_state: string;
  requested_by: UserSummary | null;
  reviewer: UserSummary | null;
}

export interface Inbox {
  requests: ReviewRequestView[];
  pending: number;
}

export function searchMembers(nickname: string): Promise<UserSummary[]> {
  if (!nickname.trim()) return Promise.resolve([]);
  return authFetch<UserSummary[]>(`/users/search?nickname=${encodeURIComponent(nickname)}`);
}

export function fetchInbox(pendingOnly = false): Promise<Inbox> {
  return authFetch<Inbox>(`/reviews/inbox?pending_only=${pendingOnly}`);
}

export function fetchSent(): Promise<ReviewRequestView[]> {
  return authFetch<ReviewRequestView[]>("/reviews/sent");
}

export function sendForReview(
  benchmarkId: string,
  reviewerNickname: string,
  stage: ReviewStage,
  comment: string,
): Promise<ReviewRequest> {
  return authFetch<ReviewRequest>(`/benchmarks/${benchmarkId}/review-requests`, {
    method: "POST",
    body: JSON.stringify({ reviewer_nickname: reviewerNickname, stage, comment }),
  });
}

export function answerReview(
  requestId: string,
  decision: "approve" | "reject",
  comment: string,
): Promise<ReviewRequestView> {
  return authFetch<ReviewRequestView>(`/reviews/${requestId}/${decision}`, {
    method: "POST",
    body: JSON.stringify({ comment }),
  });
}

export function commentOnReview(requestId: string, comment: string): Promise<ReviewRequestView> {
  return authFetch<ReviewRequestView>(`/reviews/${requestId}/comment`, {
    method: "POST",
    body: JSON.stringify({ comment }),
  });
}

export function cancelReview(requestId: string): Promise<ReviewRequestView> {
  return authFetch<ReviewRequestView>(`/reviews/${requestId}/cancel`, { method: "POST" });
}

/** The pending request that gates a stage, if there is one. */
export function pendingFor(
  requests: readonly ReviewRequestView[] | undefined,
  stage: ReviewStage,
): ReviewRequestView | undefined {
  return (requests ?? []).find(
    (view) => view.request.stage === stage && view.request.status === "pending",
  );
}

/**
 * Whether the signed-in member may press Run.
 *
 * A hint for the UI, never a decision: the backend checks ownership and
 * approval state itself on every request, and this function existing
 * does not make hiding the button a security measure.
 *
 * Only "approved" qualifies — an Approver's decision (or an admin
 * override), never the owner's own submission. Earlier states used to
 * be included here because the owner could clear the gate themselves
 * on the way to running; that self-approval path no longer exists, so
 * showing Run as available from draft/pending_approval would invite a
 * click that the backend now refuses.
 */
export function canRun(
  isOwner: boolean,
  state: string,
  requests: readonly ReviewRequestView[] | undefined,
): boolean {
  if (!isOwner) return false;
  if (pendingFor(requests, "spec")) return false;
  return state === "approved";
}

/** Whether the signed-in member may accept their own results. */
export function canAcceptResult(
  isOwner: boolean,
  state: string,
  requests: readonly ReviewRequestView[] | undefined,
): boolean {
  return isOwner && state === "pending_review" && !pendingFor(requests, "result");
}

File: apps/api/planbench_api/routers/reviews.py
"""The review inbox, and answering the requests in it.

Every route here resolves the caller from their token and compares it to
the ids stored on the request. Nothing is authorized by nickname: a
nickname is how the sender found this person, not proof of being them.
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel

from planbench_api.auth import ActiveUser
from planbench_api.dependencies import get_benchmark_service, get_review_service
from planbench_api.review import ReviewRequestView, ReviewStatus
from planbench_api.review_service import ReviewService
from planbench_api.services import BenchmarkService

router = APIRouter(prefix="/reviews", tags=["reviews"])

Reviews = Annotated[ReviewService, Depends(get_review_service)]
Benchmarks = Annotated[BenchmarkService, Depends(get_benchmark_service)]


class DecisionRequest(BaseModel):
    comment: str = ""


class CommentRequest(BaseModel):
    comment: str


class InboxResponse(BaseModel):
    requests: list[ReviewRequestView]
    pending: int


@router.get("/inbox", response_model=InboxResponse)
def inbox(reviews: Reviews, user: ActiveUser, pending_only: bool = False) -> InboxResponse:
    """Requests addressed to me. ``pending`` drives the sidebar badge."""
    return InboxResponse(
        requests=reviews.inbox(user, pending_only=pending_only),
        pending=reviews.pending_count(user),
    )


@router.get("/sent", response_model=list[ReviewRequestView])
def sent(reviews: Reviews, user: ActiveUser) -> list[ReviewRequestView]:
    return reviews.sent(user)


@router.post("/{request_id}/approve", response_model=ReviewRequestView)
def approve(
    request_id: str,
    payload: DecisionRequest,
    reviews: Reviews,
    benchmarks: Benchmarks,
    user: ActiveUser,
) -> ReviewRequestView:
    _stored, answered = benchmarks.answer_review(
        request_id, user, ReviewStatus.APPROVED, payload.comment
    )
    return reviews.view(answered)


@router.post("/{request_id}/reject", response_model=ReviewRequestView)
def reject(
    request_id: str,
    payload: DecisionRequest,
    reviews: Reviews,
    benchmarks: Benchmarks,
    user: ActiveUser,
) -> ReviewRequestView:
    _stored, answered = benchmarks.answer_review(
        request_id, user, ReviewStatus.REJECTED, payload.comment
    )
    return reviews.view(answered)


@router.post("/{request_id}/comment", response_model=ReviewRequestView)
def comment(
    request_id: str, payload: CommentRequest, reviews: Reviews, user: ActiveUser
) -> ReviewRequestView:
    """Say something without deciding; the request stays pending."""
    return reviews.view(reviews.comment(request_id, user, payload.comment))


@router.post("/{request_id}/cancel", response_model=ReviewRequestView)
def cancel(
    request_id: str, reviews: Reviews, benchmarks: Benchmarks, user: ActiveUser, request: Request
) -> ReviewRequestView:
    """Withdraw a request I sent."""
    pending = reviews.get_request(request_id)
    cancelled = benchmarks.cancel_review(pending.benchmark_id, user, request_id)
    return reviews.view(cancelled)
